### PR TITLE
Monomorphization

### LINKED ISSUE #59: Add polymorphism

### PR DESCRIPTION
**Pull request by janmasrovira**
This PR implements the translation from MicroJuvix to MonoJuvix. This closes #59.

However, there are several assumptions for this to work, all listed [here](https://github.com/heliaxdev/minijuvix/blob/8fa26ad35de576cd2de27c4c0c76e9a09ec59429/notes/Monomorphization.org). The algorithm is also outlined there.
If one of the assumptions does not hold, the translation should stop with an error. If a bogus result is returned with no error, then it is a bug.

### CODE REVIEW

File: app/Commands/MicroJuvix.hs [hunk 1]
```diff
@@ -7,13 +7,21 @@ import MiniJuvix.Prelude hiding (Doc)
...
 data MicroJuvixCommand
-  = Pretty MicroJuvixOptions
-  | TypeCheck MicroJuvixOptions
+  = Pretty MicroJuvixPrettyOptions
+  | TypeCheck MicroJuvixTypeOptions
 
-newtype MicroJuvixOptions = MicroJuvixOptions
-  { _mjuvixInputFile :: FilePath
+newtype MicroJuvixPrettyOptions = MicroJuvixPrettyOptions
+  { _mjuvixPrettyInputFile :: FilePath
```
Comment (jonaprieto) [inline app/Commands/MicroJuvix.hs:14]: mjuvix → microJuvix

File: app/Commands/MicroJuvix.hs [hunk 2]
```diff
@@ -7,13 +7,21 @@ import MiniJuvix.Prelude hiding (Doc)
...
   }
 
+data MicroJuvixTypeOptions = MicroJuvixTypeOptions
+  { _mjuvixTypeInputFile :: FilePath,
+    _mjuvixTypePrint :: Bool
```
Comment (jonaprieto) [inline app/Commands/MicroJuvix.hs:19]: mjuvixTypeInputFile → microJuvixTypeOptionsInputFile

File: app/Commands/MonoJuvix.hs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE ApplicativeDo #-}
+
+module Commands.MonoJuvix where
+
+import Commands.Extra
+import MiniJuvix.Prelude hiding (Doc)
+import Options.Applicative
+
+newtype MonoJuvixOptions = MonoJuvixOptions
+  { _monojuvixInputFile :: FilePath
```
Comment (jonaprieto) [inline app/Commands/MonoJuvix.hs:10]: monojuvixInputFile → monoJuvixOptionsInputFile

File: notes/Monomorphization.org [hunk 4]
```diff
@@ -42,13 +42,17 @@ odd A (cons _ xs) ≔ not (even A xs) ;
 -- main ≔ even Bool ..  odd Nat;
 #+end_src
 
-* Algorithm
+* Collection algorithm
+ This sections describes to algorithm to collect the list of all concrete
```
Comment (jonaprieto) [inline notes/Monomorphization.org:46]: There are two typos here.

File: src/MiniJuvix/Pipeline/EntryPoint.hs [hunk 5]
```diff
@@ -2,10 +2,14 @@ module MiniJuvix.Pipeline.EntryPoint where
...
 data EntryPoint = EntryPoint
   { _entryRoot :: FilePath,
     _entryModulePaths :: NonEmpty FilePath
   }
   deriving stock (Eq, Show)
 
 makeLenses ''EntryPoint
+
+mainModulePath :: Lens' EntryPoint FilePath
+mainModulePath = entryModulePaths . _head
```
Comment (jonaprieto) [inline src/MiniJuvix/Pipeline/EntryPoint.hs:15]: Why is this needed?

File: src/MiniJuvix/Syntax/MicroJuvix/Language.hs [hunk 6]
```diff
@@ -2,12 +2,14 @@ module MiniJuvix.Syntax.MicroJuvix.Language
...
     module MiniJuvix.Syntax.Concrete.Scoped.Name.NameKind,
     module MiniJuvix.Syntax.Concrete.Scoped.Name,
+    module MiniJuvix.Syntax.Concrete.Loc,
   )
 where
 
 import MiniJuvix.Prelude
-import MiniJuvix.Syntax.Concrete.Language (HasLoc)
-import MiniJuvix.Syntax.Concrete.Language qualified as C
+import MiniJuvix.Syntax.Backends
```
Comment (jonaprieto) [inline src/MiniJuvix/Syntax/MicroJuvix/Language.hs:10]: The import of ‘MiniJuvix.Syntax.Backends’ is redundant

File: src/MiniJuvix/Syntax/MicroJuvix/Language.hs [hunk 7]
```diff
@@ -103,10 +107,17 @@ data TypedExpression = TypedExpression
   }
   deriving stock (Show)
 
+data FunctionExpression = FunctionExpression
+  { _funExprLeft :: Expression,
```
Comment (jonaprieto) [inline src/MiniJuvix/Syntax/MicroJuvix/Language.hs:111]: funExprLeft → functionExpressionLeft

File: src/MiniJuvix/Syntax/MonoJuvix/Language/Extra.hs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+module MiniJuvix.Syntax.MonoJuvix.Language.Extra
+  ( module MiniJuvix.Syntax.MonoJuvix.Language.Extra,
+    module MiniJuvix.Syntax.MonoJuvix.Language,
+  )
+where
+
+import MiniJuvix.Syntax.MonoJuvix.Language
+
+foldApplication :: Expression -> [Expression] -> Expression
+foldApplication f args = case args of
```
Comment (jonaprieto) [inline src/MiniJuvix/Syntax/MonoJuvix/Language/Extra.hs:10]: Use` \case`; `args` is not used

File: src/MiniJuvix/Translation/AbstractToMicroJuvix.hs [hunk 9]
```diff
@@ -183,11 +183,23 @@ goIden i = case i of
   A.IdenAxiom a -> IdenAxiom (goName (a ^. A.axiomRefName))
   A.IdenInductive a -> IdenInductive (goName (a ^. A.inductiveRefName))
 
+goExpressionFunction :: forall r. A.Function -> Sem r FunctionExpression
+goExpressionFunction f = do
+  l' <- goParam (f ^. A.funParameter)
+  r' <- goExpression (f ^. A.funReturn)
+  return (FunctionExpression l' r')
+  where
+    goParam :: A.FunctionParameter -> Sem r Expression
```
Comment (jonaprieto) [inline src/MiniJuvix/Translation/AbstractToMicroJuvix.hs:192]: I don't fully understand this one. Is this case needed because in a type like `(x : A) → B,` the notation `(x : A)` is not supported?

Comment (janmasrovira) [reply src/MiniJuvix/Translation/AbstractToMicroJuvix.hs:192]: that's right

File: tests/positive/VP/PolySimpleFungibleToken.mjuvix [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+--------------------------------------------------------------------------------
+-- Integers
+--------------------------------------------------------------------------------
+
+axiom Int : Type {
+  ghc ↦ "Int";
+};
+
+axiom lt : Int → Int → BackendBool {
```
Comment (jonaprieto) [inline tests/positive/VP/PolySimpleFungibleToken.mjuvix:81]: Use the new syntax for compile blocks.

File: app/Commands/MicroJuvix.hs (hunk 2 above)
Comment (janmasrovira) [reply app/Commands/MicroJuvix.hs]: fixed by d3f134341603bf568a4c2a1eaf885769c65b8ff4

File: app/Commands/MicroJuvix.hs (hunk 1 above)
Comment (janmasrovira) [reply app/Commands/MicroJuvix.hs]: fixed by d3f134341603bf568a4c2a1eaf885769c65b8ff4

File: app/Commands/MonoJuvix.hs (hunk 3 above)
Comment (janmasrovira) [reply app/Commands/MonoJuvix.hs]: fixed by d3f134341603bf568a4c2a1eaf885769c65b8ff4

File: notes/Monomorphization.org (hunk 4 above)
Comment (janmasrovira) [reply notes/Monomorphization.org]: fixed by d3f134341603bf568a4c2a1eaf885769c65b8ff4

File: src/MiniJuvix/Syntax/MicroJuvix/Language.hs (hunk 7 above)
Comment (janmasrovira) [reply src/MiniJuvix/Syntax/MicroJuvix/Language.hs]: fixed by d3f134341603bf568a4c2a1eaf885769c65b8ff4

File: src/MiniJuvix/Pipeline/EntryPoint.hs (hunk 5 above)
Comment (janmasrovira) [reply src/MiniJuvix/Pipeline/EntryPoint.hs:15]: The entry point needs to have a main module. In the monomorphization phase, we use this module to collect the starting set of concrete type applications. As a temporary solution, I pick the first element of the list, which should be the module that the user provided to the CLI. I think it is good to encapsulate this convention in a lens because if we decide to have a specific field for in in the future, we will only need a local update to the definition of this lens

File: tests/positive/VP/PolySimpleFungibleToken.mjuvix (hunk 10 above)
Comment (janmasrovira) [reply tests/positive/VP/PolySimpleFungibleToken.mjuvix]: fixed in 9441795e865853020c749b0cc10e0f0f3cfda7bb.
I have also added [PolySimpleFungibleToken.mjuvix](https://github.com/heliaxdev/minijuvix/pull/70/files/84a4850d02358de80be2e8ecb9357e7c42390397#diff-ea297effdeac2854253d51dcec17553769813c11ab4e3df342d2c6771a08b5c2) to the test suite

File: src/MiniJuvix/Syntax/MicroJuvix/Language/Extra.hs [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+type ConcreteTypeCall = TypeCall' ConcreteType
+
+type TypeCall = TypeCall' Type
+
+type SubsE = HashMap VarName Expression
+
+type Rename = HashMap VarName VarName
+
+type Subs = HashMap VarName Type
```
Comment (jonaprieto) [inline src/MiniJuvix/Syntax/MicroJuvix/Language/Extra.hs:46]: Here and below, what does the word `Subs` mean in the names?

Comment (janmasrovira) [reply src/MiniJuvix/Syntax/MicroJuvix/Language/Extra.hs:46]: Substitution

**Review by jonaprieto — APPROVED** [review]

File: src/MiniJuvix/Syntax/MonoJuvix/Language/Extra.hs (hunk 8 above)
Comment (jonaprieto) [reply src/MiniJuvix/Syntax/MonoJuvix/Language/Extra.hs:10]: This will be listed in the future by Hlint, I hope.

Comment (janmasrovira) [reply src/MiniJuvix/Syntax/MonoJuvix/Language/Extra.hs:10]: I didn't use `\case` since by keeping the `args` name I thought it was more clear what the function does

File: tests/positive/Polymorphism2.mjuvix [hunk 12]
```diff
@@ -0,0 +1,44 @@
+module Polymorphism2;
```
Comment (paulcadman) [inline tests/positive/Polymorphism2.mjuvix:1]: Did you mean to commit this file - or is it incomplete? It's not run as part of the tests and if I run monojuvix manually I get:

```
minijuvix: the given type is not concrete
CallStack (from HasCallStack):
  error, called at src/MiniJuvix/Prelude/Base.hs:242:9 in minijuvix-0.1.2-AfIq75bwxFQB6Y2Nn2uZ0H:MiniJuvix.Prelude.Base
  error, called at src/MiniJuvix/Syntax/MicroJuvix/Language/Extra.hs:82:14 in minijuvix-0.1.2-AfIq75bwxFQB6Y2Nn2uZ0H:MiniJuvix.Syntax.MicroJuvix.Language.Extra
```

Comment (janmasrovira) [reply tests/positive/Polymorphism2.mjuvix:1]: this file slipped, my bad.
I don't immediately see what causes the error so I'll need to look into it

Comment (janmasrovira) [reply tests/positive/Polymorphism2.mjuvix]: I found the error. It is fixed in ec2809d0c0998fae4d210ad7622f416c964dd964
The problem was that if a polymorphic inductive definition was never used, it had no associated type calls and thus it was (incorrectly) assumed to be concrete.

**Review by paulcadman — APPROVED** [review]
